### PR TITLE
[ft-template] Add llm-forge cookbook: full-parameter vs LoRA

### PR DESCRIPTION
**Pull request by ArturNiederfahrenhorst**
Adds the full-parameter vs LoRA cookbook

https://docs.google.com/document/d/15A3-Ey0UHGFUQtxtvmkchBDeFvjsuVil7HD5xUS7kxU/edit

### CODE REVIEW

File: templates/fine-tune-llm_v2/cookbooks/lora_vs_full_parameter/README.md [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+
+### How to launch a LoRA fine-tuning job
+
+
+```python
+# [Optional] You can set the WandB API key to track model performance
+# !export WANDB_API_KEY={YOUR_WANDB_API_KEY}
+
+# Fine-tune Llama 3 8B with LoRA
+!python main.py llama-3-8b-lora.yaml
```
Comment (kouroshHakha) [inline templates/fine-tune-llm_v2/cookbooks/lora_vs_full_parameter/README.md:32]: we should already have lora vs. full param in the config list one level above this. So why not reuse them?

File: templates/fine-tune-llm_v2/cookbooks/lora_vs_full_parameter/README.md [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+
+There is no general answer to this but here are some things to consider:
+
+- The quality of the fine-tuned models will, in most cases, be comparable if not the same
+- LoRA shines if...
+    - ... you want to serve many fine-tuned models at once yourself
+    - ... you want to rapidly experiment (because fine-tuning, downloading and serving the model take less time)
+- Full-parameter shines if...
+    - ... you want to make sure that your fine-tuned model has the maximum quality
+    - ... you want to serve only one fine-tuned version of the model
```
Comment (kouroshHakha) [inline templates/fine-tune-llm_v2/cookbooks/lora_vs_full_parameter/README.md:96]:
```suggestion
- LoRA shines if:
    - You want to serve many fine-tuned models at once yourself
    - You want to rapidly experiment (because fine-tuning, downloading and serving the model take less time)
- Full-parameter shines if:
    - You want to make sure that your fine-tuned model has the maximum quality
    - You want to serve only one fine-tuned version of the model
```

File: templates/fine-tune-llm_v2/cookbooks/lora_vs_full_parameter/README.md [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+
+[LoRA (Low-Rank Adaptation)](https://arxiv.org/abs/2106.09685) is a fine-tuning technique that freezes all the weights of your LLM and adds a few parameters to it that get fine-tuned instead. These additional parameters make up a LoRA checkpoint. There are two important things to take away from this:
+1. Since all the original weights are frozen, they don't have to be optimized and therefore don't take up as many resources during fine-tuning. In practice, you can fine-tune on a smaller cluster.
+2. Since the checkpoint only consists of the few additional parameters, it is very small. If we load the original model into memory, we can swap out the fine-tuned weights quickly. Therefore, it makes for an efficient scheme for serving many fine-tuned models alongside each other.
+
+You can find a more in-depth analysis of this topic [here](https://www.anyscale.com/blog/fine-tuning-llms-lora-or-full-parameter-an-in-depth-analysis-with-llama-2).
```
Comment (kouroshHakha) [inline templates/fine-tune-llm_v2/cookbooks/lora_vs_full_parameter/README.md:16]: I would add some more recent research here as well:
- https://arxiv.org/abs/2405.09673

**Review by kouroshHakha — COMMENTED** [review]
some comments:

File: templates/fine-tune-llm_v2/cookbooks/lora_vs_full_parameter/README.md (hunk 1 above)
Comment (ArturNiederfahrenhorst) [reply templates/fine-tune-llm_v2/cookbooks/lora_vs_full_parameter/README.md:32]: This won't work with the deepspeed config, because that's a relative path as well.
We could manipulate that in main but I think it's wise here to keep it simple since all of this is code that users can end up reading and get confused about.

File: templates/fine-tune-llm_v2/cookbooks/lora_vs_full_parameter/README.md (hunk 3 above)
Comment (ArturNiederfahrenhorst) [reply templates/fine-tune-llm_v2/cookbooks/lora_vs_full_parameter/README.md:16]: Yeah that's the paper I had in mind the other day, only that it was not from HF 😄

File: templates/fine-tune-llm_v2/cookbooks/lora_vs_full_parameter/README.ipynb
Comment (kouroshHakha) [inline templates/fine-tune-llm_v2/cookbooks/lora_vs_full_parameter/README.ipynb:1]: Add the cook books to the main file and update the cookbooks list. Also you can remove the FAQ about lora in the main README.

File: templates/fine-tune-llm_v2/cookbooks/lora_vs_full_parameter/README.md (hunk 1 above)
Comment (kouroshHakha) [reply templates/fine-tune-llm_v2/cookbooks/lora_vs_full_parameter/README.md:34]: I made it work here. https://github.com/anyscale/templates/blob/711038265881b1188ea6ccfbacaad94bef35a973/templates/fine-tune-llm_v2/cookbooks/continue_from_checkpoint/llama-3-8b.yaml#L17

Let's adopt this pattern instead of doing `../../deepspeed_configs`. You need to make it clear that the user should run the command from the root of the template dir.

File: templates/fine-tune-llm_v2/cookbooks/lora_vs_full_parameter/llama-3-8b-full-parameter.yaml [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+valid_path: s3://air-example-data/gsm8k/test.jsonl # <-- change this to the path to your validation data. This is optional
+context_length: 512 # <-- change this to the context length you want to use
+num_devices: 16 # <-- change this to total number of GPUs that you want to use
+num_epochs: 3 # <-- change this to the number of epochs that you want to train for
+train_batch_size_per_device: 8
+eval_batch_size_per_device: 16
+learning_rate: 1e-5 # <-- For LoRA, we recommend to default to 1e-4, for full-parameter training, we recommend 1e-5
+padding: "longest" # This will pad batches to the longest sequence. Use "max_length" when profiling to profile the worst case.
+num_checkpoints_to_keep: 1
+dataset_size_scaling_factor: 10000
```
Comment (kouroshHakha) [inline templates/fine-tune-llm_v2/cookbooks/lora_vs_full_parameter/llama-3-8b-full-parameter.yaml:12]: remove?

File: templates/fine-tune-llm_v2/cookbooks/lora_vs_full_parameter/llama-3-8b-full-parameter.yaml
Comment (kouroshHakha) [inline templates/fine-tune-llm_v2/cookbooks/lora_vs_full_parameter/llama-3-8b-full-parameter.yaml:1]: I don't think we need to repeat the yaml and main.py here. Just use the same YAMLS that are used in the main README.

File: templates/fine-tune-llm_v2/cookbooks/lora_vs_full_parameter/README.ipynb
Comment (ArturNiederfahrenhorst) [reply templates/fine-tune-llm_v2/cookbooks/lora_vs_full_parameter/README.ipynb:1]: Sorry I thought I did and committed that.

File: templates/fine-tune-llm_v2/README.ipynb [hunk 5]
```diff
@@ -6,7 +6,7 @@
    "source": [
     "# Fine-tuning Llama-3, Mistral and Mixtral with Anyscale\n",
     "\n",
-    "**⏱️ Time to complete**: 2.5 hours for 7/8B models (9 hours for 13B, 25 hours for 70B)\n",
+    "**⏱️ Time to complete**: 60min for 8B model with LoRA, 2.5h with full-parameter fine-tuning\n",
```
Comment (ArturNiederfahrenhorst) [inline templates/fine-tune-llm_v2/README.ipynb:9]: I've made additional changes to the configs so that we stop at three epochs. That's enough to learn gsm8k as per my previous experiments and certainly enough to demonstrate to people that this thing works.

-> We should steer them away from running endless experiments during their "orientation phase"

File: templates/fine-tune-llm_v2/training_configs/full_param/mixtral-8x7b.yaml [hunk 6]
```diff
@@ -3,7 +3,7 @@ train_path: s3://air-example-data/gsm8k/train.jsonl # <-- change this to the pat
 valid_path: s3://air-example-data/gsm8k/test.jsonl # <-- change this to the path to your validation data. This is optional
 context_length: 512 # <-- change this to the context length you want to use
 num_devices: 32 # <-- change this to total number of GPUs that you want to use
-num_epochs: 32 # <-- change this to the number of epochs that you want to train for
```
Comment (ArturNiederfahrenhorst) [inline templates/fine-tune-llm_v2/training_configs/full_param/mixtral-8x7b.yaml:6]: This was a mistake I believe

**Comment by ArturNiederfahrenhorst** [comment]
I ran both, the LoRA and the fullparameter, line from the cookbook

**Comment by ArturNiederfahrenhorst** [comment]
Since we are not adding any now workflows here now, but just text, I don't expect that we have to go through making a video here. @akshay-anyscale ?

**Comment by ArturNiederfahrenhorst** [comment]
@kouroshHakha As per our offline convo: We don't have to do a video here so let's meeeerge! :)

**Review by kouroshHakha — APPROVED** [review]
